include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,802 @@
+     * The runtime of the construction is stored.
+     */
+    const std::chrono::duration<double> runtime{};
+    /**
+     * A quick heuristic to assess the quality of the pruning is captured by the size of the charge space of the top
+     * cluster, which depends on the charge spaces of all clusters below it. The number of pruned elements is stored
+     * here, which is computed by subtracting the size of the charge space of the top cluster from the maximum size it
+     * can have for the given simulation base.
+     */
+    const uint64_t pruned_top_level_multisets{};
+    /**
+     * The maximum size of the charge space of the top cluster, given the simulation base, can be inferred by the
+     * \"stars and bars\" combinatorial idea: the solution to this analogous problem determines the maximum amount of
+     * multisets of size \f$N\f$ (where \f$N\f$ is the number of SiDBs in the layout, and therefore in the top cluster)
+     * for the given base \f$b\f$. In particular, the analogy is as follows: any such multiset can be seen as \f$N\f$
+     * stars and \f$b - 1\f$ bars separating those stars. Then, the \f$b\f$ partitions forming from these \f$b - 1\f$
+     * separators each have a respective size, adding up to \f$N\f$. Therefore each partition is associated with an
+     * amount of one of the charge states of the multiset. Now we may compute total number of possible multisets for the
+     * top cluster as the number of combinations of \f$N\f$ stars and \f$b - 1\f$ bars. Hence this is computed with the
+     * following combinatorial formula: \f$\binom{N + b - 1}{b - 1}\f$.
+     */
+    const uint64_t maximum_top_level_multisets{};
+};
+
+namespace detail
+{
+
+template <typename Lyt>
+class ground_state_space_impl
+{
+  public:
+    explicit ground_state_space_impl(const Lyt& lyt, const uint64_t max_cluster_size_for_witness_partitioning,
+                                     const sidb_simulation_parameters& phys_params) noexcept :
+            base{phys_params.base},
+            top_cluster{to_sidb_cluster(sidb_cluster_hierarchy(lyt))},
+            clustering{get_initial_clustering(top_cluster, get_local_potential_bounds(lyt, phys_params))},
+            witness_partitioning_max_cluster_size{max_cluster_size_for_witness_partitioning},
+            mu_bounds_with_error{physical_constants::POP_STABILITY_ERR - phys_params.mu_minus,
+                                 -physical_constants::POP_STABILITY_ERR - phys_params.mu_minus,
+                                 physical_constants::POP_STABILITY_ERR - phys_params.mu_plus(),
+                                 -physical_constants::POP_STABILITY_ERR - phys_params.mu_plus()}
+    {}
+
+    ground_state_space_result run() noexcept
+    {
+        mockturtle::stopwatch<>::duration time_counter{};
+        {
+            const mockturtle::stopwatch stop{time_counter};
+
+            while (!terminate)
+            {
+                while (!update_charge_spaces())
+                    ;
```

Comment:
why this syntax here?

Reply:
ClangFormat does this. The body is specifically left empty here.

Reply:
update:  Clang-Tidy demanded braces, so now it's:
```c++
while (!update_charge_spaces())
{}
```

nifty semi-colons are not appreciated, but I understand the reasoning (ever had to figure out your code was broken because of `if(condition);`?)